### DISCUSSION TITLE
Is there a simple way to manually retry jobs?

### DISCUSSION THREAD

**Discussion by JPCanaverde** [jobrunr/jobrunr#759] [Q&A] [answered]
If some jobs exceed their retry (or are scheduled to be retried a few hours later for example), is there a way to manually retry them?

I tried manually setting them to `ENQUEUED` in the database but that didn't seem to work.

Ideally, I'm looking to programatically find all failed jobs & retry them, or retry some specific jobs by id.

**Comment by rdehuyss** [comment, marked as answer]
You can try the `StorageProvider` API.

**Reply by JPCanaverde** [reply]
@rdehuyss thanks for pointing me in the right direction!

This does the trick:

```kotlin
val job = storageProvider.getJobById(jobId)

 job.enqueue()
        
storageProvider.save(job)
```

Can also get jobs by `FAILED` state (or `SCHEDULED` state with `FAILED` in the state history) which is nice for retrying all failed jobs.

**Reply by rdehuyss** [reply]
Wouldn't it make more sense to adjust your back-off policy? When / how do you decide to retry programmatically?

**Reply by JPCanaverde** [reply]
The main use case for this is a bug being introduced and causing some/all jobs to fail, rather than a temporarily failure e.g. because something is down. Once the bug fix is live, I can manually retry those failed jobs.

JobRunr's dashboard would probably be the ideal place to do that, but I'd need to do some work to expose it with my current infrastructure. Having an endpoint I can call manually is a good enough hack for now. 😅

**Reply by avinashjeevanandham** [reply]
Although `FAILED` is considered to be a terminal state and cannot be retried.

**Reply by rdehuyss** [reply]
> Although `FAILED` is considered to be a terminal state and cannot be retried.

Says who? Why could it not be retried?
